lib/srv/alpnproxy/proxy.go

```diff
@@ -379,14 +399,49 @@ func (p *Proxy) handleConn(ctx context.Context, clientConn net.Conn) error {
 		return trace.Wrap(err)
 	}
 
+	var handlerConn net.Conn = tlsConn
+	// Check if ping is supported/required by the client.
+	if common.IsPingProtocol(common.Protocol(tlsConn.ConnectionState().NegotiatedProtocol)) {
+		handlerConn = p.handlePingConnection(ctx, tlsConn)
+	}
+
 	isDatabaseConnection, err := dbutils.IsDatabaseConnection(tlsConn.ConnectionState())
 	if err != nil {
 		p.log.WithError(err).Debug("Failed to check if connection is database connection.")
 	}
 	if isDatabaseConnection {
-		return trace.Wrap(p.handleDatabaseConnection(ctx, tlsConn, connInfo))
+		return trace.Wrap(p.handleDatabaseConnection(ctx, handlerConn, connInfo))
 	}
-	return trace.Wrap(handlerDesc.handle(ctx, tlsConn, connInfo))
+	return trace.Wrap(handlerDesc.handle(ctx, handlerConn, connInfo))
+}
+
+// handlePingConnection starts the server ping routine and returns `pingConn`.
+func (p *Proxy) handlePingConnection(ctx context.Context, conn net.Conn) net.Conn {
+	pingConn := newPingConn(conn)
+
+	// Start ping routine. It will continuously send pings in a defined
+	// interval.
+	go func() {
+		ticker := time.NewTicker(p.cfg.PingInterval)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ctx.Done():
```

Comment:
If we're "done" here, shouldn't we also close the connection?

Reply:
I was following the ["recommendation" from the `Serve` function to let the handler close the connection](https://github.com/gravitational/teleport/blob/master/lib/srv/alpnproxy/proxy.go#L310). Not sure if we can have some issue with database connections tho, but I'm assuming the handlers are closing it properly.

Reply:
`ctx` here is the server context. Seems with this implementation, we are mostly relying on `pingConn.WritePing()` to fail when the connection is closed. Not sure if it is more convenient to built-in this periodic ping function inside `pingConn` so the `pingConn` can track Close better and stop the ticker.

Reply:
It might be redundant to have `ctx.Done` in this situation. Since the connection will also be closed, I can drop it too.

> Not sure if it is more convenient to built-in this periodic ping function inside `pingConn` so the `pingConn` can track Close better and stop the ticker.

I think it is better to let this outside the "protocol implementation", allowing the callers/users to decide how they will use the `WritePing` function. For example, the Local Proxy doesn't send any ping information and uses the connection.